Documentation/Profiling/ReJIT on Attach.md

```diff
@@ -0,0 +1,68 @@
+
+    // This flag controls whether the runtime will call GetReJITParameters
+    // on methods that are ReJITted because they inline a method that was requested
+    // for ReJIT
+    COR_PRF_REJIT_INLINING_CALLBACKS    = 0x2
+} COR_PRF_REJIT_FLAGS;
+```
+
+Any callers of this API must set `COR_PRF_REJIT_BLOCK_INLINING`. Although it is possible that in the future this restriction will be lifted, the current implementation blocks ReJITted methods from being inlined (ever).
+
+The other value `COR_PRF_REJIT_INLINING_CALLBACKS` controls whether you get a `ICorProfilerCallback4::GetReJITParameters` callback for any methods that are ReJITted as inliners of the requested method. The default is to not receive callbacks for these methods. You will always receive a `GetReJITParameters` callback for any methods that are explictly requested.
+
+
+## Inner workings/Limitations
+
+With this API you are no longer required to monitor JIT callbacks to manually block inlining from occurring. To acheive that the runtime now globally blocks a ReJITted method from being inlined (even if it was ReJITted with `ICorProfilerInfo4::RequestReJIT` and not the new API). Once a method is reverted with `ICorProfilerInfo4::RequestRevert` inlining will occur again for any future jittings.
+
+It is important to mention here how `RequestRevert` works. When you revert a ReJITted method, the original native code is activated. This means there are potential pitfalls for calling `RequestRevert`. Consider an app where method A has previously inlined both method B and method C (among potentially others), then a profiler calls `RequestReJITWithInliners` for B and C. If the profiler then wants to revert the ReJIT for method B, it might seem logical to revert method B and method A. However, when the method A code is reverted the original native code (without modification) will be activated, and that original code includes method C inlined. The result is that even though method C has been ReJITted, any future calls to method A will not call the ReJITted method C and will instead call the inlined (default IL) of method C.
+
+The limiation of collectible and dynamic methods has not been lifted. It is not currently possible to ReJIT these types of methods, although we would like to lift that restriction in the future.
+
+## Metadata Changes on Attach
+
+Usually profiler authors do not want to trivially change the IL for ReJITted methods, but rather inject new types/methods and call those new types/methods. Previously our guidance has been to do any metadata rewriting during the `ICorProfilerCallback::ModuleLoadFinished` callback. This advice presents a challenge if the profiler is not attached during module load and still wants to modify metadata.
+
+To work around this a set of metadata changes is now legal to make at any point as long as you call `ICorProfilerInfo7::ApplyMetadata` afterwards:
+* `DefineUserString`
+* `DefineTypeRefByName`
+* `DefineMemberRef`
+* `DefineTypeDef `
```

Comment:
Any restrictions on generic types, deriving from pre-existing types, adding new inner types to pre-existing types, abstract types?

(FWIW my nieve guess if they work would be no, yes, yes, yes but I also bet all of them are untested : )

Reply:
Generic types are fine to add, adding new inner types would be ok for new types only, and I don't know about deriving from pre-existing types or abstract types. I updated the docs to reflect my knowledge, I don't currently have plans to test the whole metadata surface area (it didn't sound like you were asking, I just wanted to be clear about my intentions)